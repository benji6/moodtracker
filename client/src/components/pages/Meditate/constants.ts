export const SEARCH_PARAM_TIME_KEY = "t";
