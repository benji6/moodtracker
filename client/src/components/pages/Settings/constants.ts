export const NEW_EMAIL_SEARCH_PARAM_KEY = "new-email";
