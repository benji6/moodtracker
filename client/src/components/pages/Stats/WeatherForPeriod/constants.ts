export const MINIMUM_LOCATION_COUNT_FOR_MEAN_CHARTS = 10;
