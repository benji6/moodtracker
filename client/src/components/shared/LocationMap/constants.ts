export const LOCATION_MAP_HEIGHT = 600;
export const LOCATION_MAP_SCALE = 147;
export const LOCATION_MAP_WIDTH = 800;
