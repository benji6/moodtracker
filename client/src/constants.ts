import addDays from "date-fns/addDays";
import startOfWeek from "date-fns/startOfWeek";
import {
  weekdayNarrowFormatter,
  weekdayShortFormatter,
  WEEK_OPTIONS,
} from "./formatters";

export const DESCRIPTION_MAX_LENGTH = 32;
export const MEDITATION_SEARCH_PARAM_TIME_KEY = "t";
export const MOOD_RANGE: [number, number] = [0, 10];
export const MOOD_INTEGERS = [
  ...Array(MOOD_RANGE[1] - MOOD_RANGE[0] + 1).keys(),
] as const;

// extracted from cognito and updated manually for now
// if needed we can look at automating this in the future
export const TOTAL_USERS = 50;

export const ERRORS = {
  integer: "Please type a valid whole number",
  network: "Something went wrong, check your internet connection and try again",
  specialCharacters: "This field must not contain any special characters",
  required: "Required",
} as const;

export const PATTERNS = {
  noPunctuation:
    "[^‒–—―|$&~=\\/⁄@+*!?({[\\]})<>‹›«».;:^‘’“”'\",،、`·•†‡°″¡¿※#№÷×%‰−‱¶′‴§_‖¦]*",
} as const;

export const FIELDS = {
  description: {
    autoComplete: "on",
    label: "Mood tags",
    maxLength: DESCRIPTION_MAX_LENGTH,
    name: "description",
    optional: true,
    pattern: PATTERNS.noPunctuation,
    supportiveText: `Add one or more words separated by spaces to describe your mood (${DESCRIPTION_MAX_LENGTH} characters max). These words will be used in your word clouds and as search terms.`,
  },
  exploration: {
    label: "Exploration",
    name: "exploration",
    optional: true,
    rows: 5,
    supportiveText:
      "Use this space to explore how you're feeling, why you're feeling that way and what's going on in your life right now",
  },
  mood: {
    label: "Mood",
    name: "mood",
  },
} as const;

export const TIME = {
  daysPerWeek: 7,
  hoursPerDay: 24,
  secondsPerMinute: 60,
} as const;

export const TEST_IDS = (() => {
  const keys = [
    "addMoodPage",
    "addMoodRadioButton",
    "addMoodSubmitButton",
    "descriptionInput",
    "meditatePage",
    "meditationCustomTimeInput",
    "meditationPresetTimeButton",
    "meditationTimerPage",
    "moodCardMood",
    "moodCardTime",
    "moodList",
    "navButton",
    "resetPasswordPage",
    "signInLink",
    "signOutButton",
    "signOutConfirmButton",
    "statsOverviewPage",
  ] as const;
  const testIds = {} as {
    [k in typeof keys[number]]: typeof keys[number];
  };
  for (const key of keys) testIds[key] = key;
  return testIds;
})();

const now = Date.now();
const startOfWeekDate = startOfWeek(now, WEEK_OPTIONS);

const weekdayDates = [
  addDays(startOfWeekDate, 0),
  addDays(startOfWeekDate, 1),
  addDays(startOfWeekDate, 2),
  addDays(startOfWeekDate, 3),
  addDays(startOfWeekDate, 4),
  addDays(startOfWeekDate, 5),
  addDays(startOfWeekDate, 6),
] as const;

type WeekdayLabels = [string, string, string, string, string, string, string];
export const WEEKDAY_LABELS_NARROW: WeekdayLabels = weekdayDates.map((date) =>
  weekdayNarrowFormatter.format(date)
) as WeekdayLabels;

export const WEEKDAY_LABELS_SHORT: WeekdayLabels = weekdayDates.map((date) =>
  weekdayShortFormatter.format(date)
) as WeekdayLabels;
