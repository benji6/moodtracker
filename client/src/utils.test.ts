import {
  getEnvelopingMoodIds,
  moodToColor,
  trapeziumArea,
  mapRight,
  computeAverageMoodInInterval,
  roundDateDown,
  roundDateUp,
  formatIsoMonthInLocalTimezone,
  getMoodIdsInInterval,
  formatIsoDateInLocalTimezone,
  computeStandardDeviation,
  computeMean,
  isoDateFromIsoDateAndTime,
  getWeekdayIndex,
  formatIsoYearInLocalTimezone,
  createDateFromLocalDateString,
  getNormalizedDescriptionWordsFromMood,
} from "./utils";
import { MOOD_RANGE } from "./constants";

describe("utils", () => {
  describe("computeAverageMoodInInterval", () => {
    describe("when the fromDate is after the toDate", () => {
      it("returns undefined", () => {
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-31T00:00:00.000Z"),
            new Date("2020-07-30T00:00:00.000Z")
          )
        ).toBeUndefined();
      });
    });

    describe("when there are 0 moods", () => {
      it("returns undefined", () => {
        expect(
          computeAverageMoodInInterval(
            {
              allIds: [],
              byId: {},
            },
            new Date("2020-07-30T00:00:00.000Z"),
            new Date("2020-07-31T00:00:00.000Z")
          )
        ).toBeUndefined();
      });
    });

    describe("when there is 1 mood", () => {
      it("returns an average mood when the mood intersects with the interval", () => {
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-28T00:00:00.000Z"),
            new Date("2020-07-28T00:00:00.000Z")
          )
        ).toEqual(5);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-27T00:00:00.000Z"),
            new Date("2020-07-28T00:00:00.000Z")
          )
        ).toEqual(5);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-28T00:00:00.000Z"),
            new Date("2020-07-29T00:00:00.000Z")
          )
        ).toEqual(5);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-27T00:00:00.000Z"),
            new Date("2020-07-29T00:00:00.000Z")
          )
        ).toEqual(5);
      });

      it("returns undefined when the mood does not intersect with the interval", () => {
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-25T00:00:00.000Z"),
            new Date("2020-07-25T00:00:00.000Z")
          )
        ).toBeUndefined();
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-24T00:00:00.000Z"),
            new Date("2020-07-25T00:00:00.000Z")
          )
        ).toBeUndefined();
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-30T00:00:00.000Z"),
            new Date("2020-07-30T00:00:00.000Z")
          )
        ).toBeUndefined();
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z"],
              byId: { "2020-07-28T00:00:00.000Z": { mood: 5 } },
            },
            new Date("2020-07-30T00:00:00.000Z"),
            new Date("2020-07-31T00:00:00.000Z")
          )
        ).toBeUndefined();
      });
    });

    describe("when there are multiple moods", () => {
      it("works with 2 moods in the interval", () => {
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z", "2020-07-29T00:00:00.000Z"],
              byId: {
                "2020-07-28T00:00:00.000Z": { mood: 5 },
                "2020-07-29T00:00:00.000Z": { mood: 7 },
              },
            },
            new Date("2020-07-28T00:00:00.000Z"),
            new Date("2020-07-29T00:00:00.000Z")
          )
        ).toEqual(6);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z", "2020-07-29T00:00:00.000Z"],
              byId: {
                "2020-07-28T00:00:00.000Z": { mood: 5 },
                "2020-07-29T00:00:00.000Z": { mood: 7 },
              },
            },
            new Date("2020-07-27T00:00:00.000Z"),
            new Date("2020-07-29T00:00:00.000Z")
          )
        ).toEqual(6);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z", "2020-07-29T00:00:00.000Z"],
              byId: {
                "2020-07-28T00:00:00.000Z": { mood: 5 },
                "2020-07-29T00:00:00.000Z": { mood: 7 },
              },
            },
            new Date("2020-07-28T00:00:00.000Z"),
            new Date("2020-07-30T00:00:00.000Z")
          )
        ).toEqual(6);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-28T00:00:00.000Z", "2020-07-29T00:00:00.000Z"],
              byId: {
                "2020-07-28T00:00:00.000Z": { mood: 5 },
                "2020-07-29T00:00:00.000Z": { mood: 7 },
              },
            },
            new Date("2020-07-27T00:00:00.000Z"),
            new Date("2020-08-02T00:00:00.000Z")
          )
        ).toEqual(6);
      });

      it("works with 2 moods and only one in the interval", () => {
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-24T00:00:00.000Z", "2020-07-28T00:00:00.000Z"],
              byId: {
                "2020-07-24T00:00:00.000Z": { mood: 4 },
                "2020-07-28T00:00:00.000Z": { mood: 5 },
              },
            },
            new Date("2020-07-26T00:00:00.000Z"),
            new Date("2020-08-02T00:00:00.000Z")
          )
        ).toEqual(4.75);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-24T00:00:00.000Z", "2020-07-28T00:00:00.000Z"],
              byId: {
                "2020-07-24T00:00:00.000Z": { mood: 4 },
                "2020-07-28T00:00:00.000Z": { mood: 5 },
              },
            },
            new Date("2020-07-23T00:00:00.000Z"),
            new Date("2020-07-24T00:00:00.000Z")
          )
        ).toEqual(4);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-24T00:00:00.000Z", "2020-07-28T00:00:00.000Z"],
              byId: {
                "2020-07-24T00:00:00.000Z": { mood: 4 },
                "2020-07-28T00:00:00.000Z": { mood: 5 },
              },
            },
            new Date("2020-07-28T00:00:00.000Z"),
            new Date("2020-08-02T00:00:00.000Z")
          )
        ).toEqual(5);
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-24T00:00:00.000Z", "2020-07-28T00:00:00.000Z"],
              byId: {
                "2020-07-24T00:00:00.000Z": { mood: 4 },
                "2020-07-28T00:00:00.000Z": { mood: 5 },
              },
            },
            new Date("2020-07-20T00:00:00.000Z"),
            new Date("2020-07-25T00:00:00.000Z")
          )
        ).toEqual(4.125);
      });

      it("works with 2 moods and both outside the interval", () => {
        expect(
          computeAverageMoodInInterval(
            {
              allIds: ["2020-07-24T00:00:00.000Z", "2020-07-28T00:00:00.000Z"],
              byId: {
                "2020-07-24T00:00:00.000Z": { mood: 4 },
                "2020-07-28T00:00:00.000Z": { mood: 5 },
              },
            },
            new Date("2020-07-25T00:00:00.000Z"),
            new Date("2020-07-27T00:00:00.000Z")
          )
        ).toEqual(4.5);
      });
    });
  });

  test("computeMean", () => {
    expect(computeMean([])).toBeUndefined();
    expect(computeMean([5])).toBe(5);
    expect(computeMean([1, 5])).toBe(3);
    expect(computeMean([1, 2, 3, 4, 5, 6, 7])).toBe(4);
  });

  test("createDateFromLocalDateString", () => {
    expect(createDateFromLocalDateString("2021-01-01")).toEqual(
      new Date("2021-01-01T00:00:00")
    );
  });

  describe("computeStandardDeviation", () => {
    it("returns 0 when the array is empty", () => {
      expect(computeStandardDeviation([])).toBe(0);
    });

    it("returns 0 when the array contains a single item", () => {
      expect(computeStandardDeviation([5])).toBe(0);
    });

    it("returns 0 when all items in the array are the same", () => {
      expect(computeStandardDeviation([5, 5, 5, 5, 5])).toBe(0);
    });

    it("returns the correct standard deviation", () => {
      expect(
        computeStandardDeviation([727.7, 1086.5, 1091, 1361.3, 1490.5, 1956.1])
      ).toBeCloseTo(420.96);
    });
  });

  describe("getEnvelopingMoodIds", () => {
    it("throws an error when the fromDate is after the toDate", () => {
      expect(() =>
        getEnvelopingMoodIds(
          [],
          new Date("2020-09-01T00:00:00"),
          new Date("2020-09-01T00:00:00")
        )
      ).not.toThrow();
      expect(() =>
        getEnvelopingMoodIds(
          [],
          new Date("2020-09-01T00:00:01"),
          new Date("2020-09-01T00:00:00")
        )
      ).toThrow(Error("`fromDate` should not be after `toDate`"));
    });

    it("returns the first ID when the range is before the mood ID range", () => {
      const allIds = [
        "2020-10-04T00:00:00",
        "2020-10-05T00:00:00",
        "2020-10-06T00:00:00",
      ];
      expect(
        getEnvelopingMoodIds(
          allIds,
          new Date("2020-09-01T00:00:00"),
          new Date("2020-09-02T00:00:00")
        )
      ).toEqual(["2020-10-04T00:00:00"]);
    });

    it("returns the last ID when the range is after the mood ID range", () => {
      const allIds = [
        "2020-10-04T00:00:00",
        "2020-10-05T00:00:00",
        "2020-10-06T00:00:00",
      ];
      expect(
        getEnvelopingMoodIds(
          allIds,
          new Date("2020-11-01T00:00:00"),
          new Date("2020-11-02T00:00:00")
        )
      ).toEqual(["2020-10-06T00:00:00"]);
    });

    it("returns all IDs when the date range encompasses all IDs", () => {
      const allIds = [
        "2020-10-04T00:00:00",
        "2020-10-05T00:00:00",
        "2020-10-06T00:00:00",
      ];
      expect(
        getEnvelopingMoodIds(
          allIds,
          new Date("2020-10-01T00:00:00"),
          new Date("2020-10-31T00:00:00")
        )
      ).toEqual(allIds);
    });

    it("returns all IDs when the date range is equal to the mood IDs range", () => {
      const allIds = [
        "2020-10-04T00:00:00",
        "2020-10-05T00:00:00",
        "2020-10-06T00:00:00",
      ];
      expect(
        getEnvelopingMoodIds(
          allIds,
          new Date("2020-10-04T00:00:00"),
          new Date("2020-10-06T00:00:00")
        )
      ).toEqual(allIds);
    });

    it("returns first ID before range", () => {
      const allIds = [
        "2020-10-04T00:00:00.000Z",
        "2020-10-04T00:00:01.000Z",
        "2020-10-04T00:00:02.000Z",
        "2020-10-05T00:00:00.000Z",
      ];
      expect(
        getEnvelopingMoodIds(
          allIds,
          new Date("2020-10-04T00:00:02.000Z"),
          new Date("2020-10-06T00:00:00.000Z")
        )
      ).toEqual([
        "2020-10-04T00:00:01.000Z",
        "2020-10-04T00:00:02.000Z",
        "2020-10-05T00:00:00.000Z",
      ]);
    });

    it("returns first ID after range", () => {
      const allIds = [
        "2020-10-04T00:00:00.000Z",
        "2020-10-05T00:00:00.000Z",
        "2020-10-06T00:00:00.000Z",
        "2020-10-06T00:00:01.000Z",
        "2020-10-06T00:00:02.000Z",
      ];
      expect(
        getEnvelopingMoodIds(
          allIds,
          new Date("2020-10-04T00:00:00.000Z"),
          new Date("2020-10-06T00:00:00.000Z")
        )
      ).toEqual([
        "2020-10-04T00:00:00.000Z",
        "2020-10-05T00:00:00.000Z",
        "2020-10-06T00:00:00.000Z",
        "2020-10-06T00:00:01.000Z",
      ]);
    });

    it("returns first ID before range and first ID after range", () => {
      const allIds = [
        "2020-10-04T00:00:00.000Z",
        "2020-10-04T00:00:01.000Z",
        "2020-10-04T00:00:02.000Z",
        "2020-10-05T00:00:00.000Z",
        "2020-10-06T00:00:00.000Z",
        "2020-10-06T00:00:01.000Z",
        "2020-10-06T00:00:02.000Z",
      ];
      expect(
        getEnvelopingMoodIds(
          allIds,
          new Date("2020-10-04T00:00:02.000Z"),
          new Date("2020-10-06T00:00:00.000Z")
        )
      ).toEqual([
        "2020-10-04T00:00:01.000Z",
        "2020-10-04T00:00:02.000Z",
        "2020-10-05T00:00:00.000Z",
        "2020-10-06T00:00:00.000Z",
        "2020-10-06T00:00:01.000Z",
      ]);
    });
  });

  describe("getMoodIdsInInterval", () => {
    it("throws an error when the fromDate is after the toDate", () => {
      expect(() =>
        getMoodIdsInInterval(
          [],
          new Date("2020-09-01T00:00:00"),
          new Date("2020-09-01T00:00:00")
        )
      ).not.toThrow();
      expect(() =>
        getMoodIdsInInterval(
          [],
          new Date("2020-09-01T00:00:01"),
          new Date("2020-09-01T00:00:00")
        )
      ).toThrow(Error("`fromDate` should not be after `toDate`"));
    });

    it("returns an empty array when there are no mood IDs provided", () => {
      expect(
        getMoodIdsInInterval(
          [],
          new Date("2020-09-02T00:00:00"),
          new Date("2020-09-03T00:00:00")
        )
      ).toEqual([]);
    });

    it("returns an empty array when there are no moods within the interval", () => {
      expect(
        getMoodIdsInInterval(
          ["2020-09-01T23:59:59", "2020-09-03T00:00:01"],
          new Date("2020-09-02T00:00:00"),
          new Date("2020-09-03T00:00:00")
        )
      ).toEqual([]);
    });

    it("returns all moods when all moods are within the interval", () => {
      expect(
        getMoodIdsInInterval(
          ["2020-09-02T00:00:00", "2020-09-03T00:00:00"],
          new Date("2020-09-02T00:00:00"),
          new Date("2020-09-03T00:00:00")
        )
      ).toEqual(["2020-09-02T00:00:00", "2020-09-03T00:00:00"]);
    });

    it("only returns moods that are within the interval", () => {
      expect(
        getMoodIdsInInterval(
          [
            "2020-09-01T23:59:59",
            "2020-09-02T00:00:00",
            "2020-09-03T00:00:00",
            "2020-09-03T00:00:01",
          ],
          new Date("2020-09-02T00:00:00"),
          new Date("2020-09-03T00:00:00")
        )
      ).toEqual(["2020-09-02T00:00:00", "2020-09-03T00:00:00"]);
    });
  });

  test("getNormalizedDescriptionWordsFromMood", () => {
    expect(getNormalizedDescriptionWordsFromMood({ mood: 5 })).toEqual([]);
    expect(
      getNormalizedDescriptionWordsFromMood({ description: "", mood: 5 })
    ).toEqual([]);
    expect(
      getNormalizedDescriptionWordsFromMood({ description: "   ", mood: 5 })
    ).toEqual([]);
    expect(
      getNormalizedDescriptionWordsFromMood({ description: "pikachu", mood: 5 })
    ).toEqual(["Pikachu"]);
    expect(
      getNormalizedDescriptionWordsFromMood({
        description: "   pikachu   ",
        mood: 5,
      })
    ).toEqual(["Pikachu"]);
    expect(
      getNormalizedDescriptionWordsFromMood({
        description: "  Bulbasaur pIkaChu  🙂   ",
        mood: 5,
      })
    ).toEqual(["Bulbasaur", "Pikachu", "🙂"]);
  });

  test("formatIsoDateInLocalTimezone", () => {
    expect(formatIsoDateInLocalTimezone(new Date("2020-09-01T00:00:00"))).toBe(
      "2020-09-01"
    );
    expect(formatIsoDateInLocalTimezone(new Date("2020-09-30T23:59:59"))).toBe(
      "2020-09-30"
    );
    expect(formatIsoDateInLocalTimezone(new Date("2020-10-01T00:00:01"))).toBe(
      "2020-10-01"
    );
  });

  test("formatIsoMonthInLocalTimezone", () => {
    expect(formatIsoMonthInLocalTimezone(new Date("2020-09-01T00:00:00"))).toBe(
      "2020-09"
    );
    expect(formatIsoMonthInLocalTimezone(new Date("2020-09-30T23:59:59"))).toBe(
      "2020-09"
    );
    expect(formatIsoMonthInLocalTimezone(new Date("2020-10-01T00:00:00"))).toBe(
      "2020-10"
    );
  });

  test("formatIsoYearInLocalTimezone", () => {
    expect(formatIsoYearInLocalTimezone(new Date("2020-01-01T00:00:00"))).toBe(
      "2020"
    );
    expect(formatIsoYearInLocalTimezone(new Date("2020-12-31T23:59:59"))).toBe(
      "2020"
    );
    expect(formatIsoYearInLocalTimezone(new Date("2021-01-01T00:00:00"))).toBe(
      "2021"
    );
  });

  test("getWeekdayIndex", () => {
    expect(getWeekdayIndex(new Date("2020-09-01T00:00:00"))).toBe(1);
    expect(getWeekdayIndex(new Date("2020-09-02T00:00:00"))).toBe(2);
    expect(getWeekdayIndex(new Date("2020-09-03T00:00:00"))).toBe(3);
    expect(getWeekdayIndex(new Date("2020-09-04T00:00:00"))).toBe(4);
    expect(getWeekdayIndex(new Date("2020-09-05T00:00:00"))).toBe(5);
    expect(getWeekdayIndex(new Date("2020-09-06T00:00:00"))).toBe(6);
    expect(getWeekdayIndex(new Date("2020-09-07T00:00:00"))).toBe(0);
    expect(getWeekdayIndex(new Date("2020-09-08T00:00:00"))).toBe(1);
  });

  test("isoDateFromIsoDateAndTime", () => {
    expect(isoDateFromIsoDateAndTime("2020-09-01T00:00:00")).toBe("2020-09-01");
    expect(isoDateFromIsoDateAndTime("2020-09-01T23:59:59")).toBe("2020-09-01");
  });

  test("mapRight", () => {
    expect(mapRight([], (x) => x + 1)).toEqual([]);
    expect(mapRight([1], (x) => x + 1)).toEqual([2]);
    expect(mapRight([1, 2, 3], (x) => x + 1)).toEqual([4, 3, 2]);
  });

  test("moodToColor", () => {
    expect(moodToColor(MOOD_RANGE[0])).toMatchInlineSnapshot(
      `"rgb(23, 71, 240)"`
    );
    expect(
      moodToColor(MOOD_RANGE[0] + (MOOD_RANGE[1] + MOOD_RANGE[0]) / 2)
    ).toMatchInlineSnapshot(`"rgb(0, 224, 224)"`);
    expect(moodToColor(MOOD_RANGE[1])).toMatchInlineSnapshot(
      `"rgb(48, 255, 32)"`
    );
    expect(
      moodToColor(MOOD_RANGE[0] + (MOOD_RANGE[1] - MOOD_RANGE[0]) / Math.PI)
    ).toMatchInlineSnapshot(`"rgb(0, 181, 250)"`);
  });

  test("roundDateDown", () => {
    expect(roundDateDown(new Date("2020-09-09T00:00:00.000"))).toEqual(
      new Date("2020-09-09T00:00:00.000")
    );
    expect(roundDateDown(new Date("2020-09-09T00:00:00.001"))).toEqual(
      new Date("2020-09-09T00:00:00.000")
    );
    expect(roundDateDown(new Date("2020-09-09T23:59:59.999"))).toEqual(
      new Date("2020-09-09T00:00:00.000")
    );
  });

  test("roundDateUp", () => {
    expect(roundDateUp(new Date("2020-09-08T00:00:00.000"))).toEqual(
      new Date("2020-09-08T00:00:00.000")
    );
    expect(roundDateUp(new Date("2020-09-08T00:00:00.001"))).toEqual(
      new Date("2020-09-09T00:00:00.000")
    );
    expect(roundDateUp(new Date("2020-09-07T23:59:59.999"))).toEqual(
      new Date("2020-09-08T00:00:00.000")
    );
  });

  test("trapeziumArea", () => {
    expect(trapeziumArea(1, 2, 3)).toBe(4.5);
    expect(trapeziumArea(2, 4, 5)).toBe(15);
    expect(trapeziumArea(3, 3, 3)).toBe(9);
  });
});
