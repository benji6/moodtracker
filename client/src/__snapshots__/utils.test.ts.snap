// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 200}) 1`] = `
{
  "iconName": "lightning",
  "weatherColor": "var(--color-figure-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 232}) 1`] = `
{
  "iconName": "lightning",
  "weatherColor": "var(--color-figure-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 300}) 1`] = `
{
  "iconName": "drizzle",
  "weatherColor": "steelblue",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 321}) 1`] = `
{
  "iconName": "drizzle",
  "weatherColor": "steelblue",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 500}) 1`] = `
{
  "iconName": "rain",
  "weatherColor": "blue",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 531}) 1`] = `
{
  "iconName": "rain",
  "weatherColor": "blue",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 600}) 1`] = `
{
  "iconName": "snow",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 622}) 1`] = `
{
  "iconName": "snow",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 701}) 1`] = `
{
  "iconName": "menu",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 762}) 1`] = `
{
  "iconName": "menu",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 771}) 1`] = `
{
  "iconName": "wind",
  "weatherColor": "var(--color-figure-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 781}) 1`] = `
{
  "iconName": "wind",
  "weatherColor": "var(--color-figure-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 800}) 1`] = `
{
  "iconName": "moon",
  "weatherColor": "rebeccapurple",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 801}) 1`] = `
{
  "iconName": "partly-cloudy-night",
  "weatherColor": "rebeccapurple",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 802}) 1`] = `
{
  "iconName": "partly-cloudy-night",
  "weatherColor": "rebeccapurple",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 803}) 1`] = `
{
  "iconName": "cloud",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": false, "weatherId": 804}) 1`] = `
{
  "iconName": "cloud",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 200}) 1`] = `
{
  "iconName": "lightning",
  "weatherColor": "var(--color-figure-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 232}) 1`] = `
{
  "iconName": "lightning",
  "weatherColor": "var(--color-figure-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 300}) 1`] = `
{
  "iconName": "drizzle",
  "weatherColor": "steelblue",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 321}) 1`] = `
{
  "iconName": "drizzle",
  "weatherColor": "steelblue",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 500}) 1`] = `
{
  "iconName": "rain",
  "weatherColor": "blue",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 531}) 1`] = `
{
  "iconName": "rain",
  "weatherColor": "blue",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 600}) 1`] = `
{
  "iconName": "snow",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 622}) 1`] = `
{
  "iconName": "snow",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 701}) 1`] = `
{
  "iconName": "menu",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 762}) 1`] = `
{
  "iconName": "menu",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 771}) 1`] = `
{
  "iconName": "wind",
  "weatherColor": "var(--color-figure-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 781}) 1`] = `
{
  "iconName": "wind",
  "weatherColor": "var(--color-figure-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 800}) 1`] = `
{
  "iconName": "sun",
  "weatherColor": "orange",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 801}) 1`] = `
{
  "iconName": "partly-cloudy-day",
  "weatherColor": "orange",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 802}) 1`] = `
{
  "iconName": "partly-cloudy-day",
  "weatherColor": "orange",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 803}) 1`] = `
{
  "iconName": "cloud",
  "weatherColor": "var(--color-balance-more)",
}
`;

exports[`utils getWeatherIconAndColor({"isDaytime": true, "weatherId": 804}) 1`] = `
{
  "iconName": "cloud",
  "weatherColor": "var(--color-balance-more)",
}
`;
